livekit-agents/livekit/agents/utils/pool.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import time
+from typing import Awaitable, Callable, Generic, Optional, Set, TypeVar
+
+T = TypeVar("T")
+
+logger = logging.getLogger("livekit.agents.pool")
+
+
+class ConnectionPool(Generic[T]):
+    """Helper class to manage persistent connections like websockets.
+
+    Handles connection pooling and reconnection after max duration.
+    """
+
+    def __init__(
+        self,
+        *,
+        max_session_duration: Optional[float] = None,
+        connect_cb: Optional[Callable[[], Awaitable[T]]] = None,
+        close_cb: Optional[Callable[[T], Awaitable[None]]] = None,
+    ) -> None:
+        """Initialize the connection wrapper.
+
+        Args:
+            max_session_duration: Maximum duration in seconds before forcing reconnection
+            connect_cb: Optional async callback to create new connections
+            close_cb: Optional async callback to close connections
+        """
+        self._max_session_duration = max_session_duration
+        self._connect_cb = connect_cb
+        self._close_cb = close_cb
+        self._connections: dict[T, float] = {}  # conn -> connected_at timestamp
+        self._available: Set[T] = set()  # Available/reusable connections
+
+        # New set to store connections to be reaped (closed) later.
+        self._to_close: Set[T] = set()
+
+    async def connect(self) -> T:
+        """Create a new connection.
+
+        Returns:
+            The new connection object
+
+        Raises:
+            NotImplementedError: If no connect callback was provided
+        """
+        if self._connect_cb is None:
+            raise NotImplementedError("Must provide connect_cb or implement connect()")
+        return await self._connect_cb()
+
+    async def _drain_to_close(self) -> None:
+        """Drain and close all the connections queued for closing."""
+        for conn in list(self._to_close):
+            await self._maybe_close_connection(conn)
+        self._to_close.clear()
+
+    async def get(self) -> T:
+        """Get an available connection or create a new one if needed.
+
+        Returns:
+            An active connection object
+        """
+        # First, close out any connections that have previously been marked.
+        await self._drain_to_close()
```

Comment:
should this be handled inside get()?

Reply:
I didn't want to make remove/reset/etc worry about being async. it will slow down get a bit.. but if any previous connections have been removed, the next time we interact with get would clean it up